p11-kit/generate-keypair.c

```diff
@@ -0,0 +1,467 @@
+		m->mechanism = CKM_EC_EDWARDS_KEY_PAIR_GEN;
+	else
+		return false;
+
+	m->pParameter = NULL_PTR;
+	m->ulParameterLen = 0;
+
+	return true;
+}
+
+static bool
+get_ec_params (const char *curve,
+	       p11_buffer *ec_params)
+{
+	static const uint8_t OID_SECP256R1[] = { 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07 };
+	static const uint8_t OID_SECP384R1[] = { 0x06, 0x05, 0x2b, 0x81, 0x04, 0x00, 0x22 };
+	static const uint8_t OID_SECP521R1[] = { 0x06, 0x05, 0x2b, 0x81, 0x04, 0x00, 0x23 };
+	const uint8_t *data = NULL;
+	size_t len = 0;
+	bool ok = true;
+
+	if (p11_ascii_strcaseeq (curve, "secp256r1")) {
+		data = OID_SECP256R1;
+		len = sizeof (OID_SECP256R1);
+	} else if (p11_ascii_strcaseeq (curve, "secp384r1")) {
+		data = OID_SECP384R1;
+		len = sizeof (OID_SECP384R1);
+	} else if (p11_ascii_strcaseeq (curve, "secp521r1")) {
+		data = OID_SECP521R1;
+		len = sizeof (OID_SECP521R1);
+	} else
+		ok = false;
+
+	p11_buffer_init_full (ec_params, (void *)data, len, 0, NULL, NULL);
+	return ok;
+}
+
+static bool
+check_args (CK_MECHANISM_TYPE type,
+	    CK_ULONG bits,
+	    const p11_buffer *ec_params)
+{
+	if (type == CKA_INVALID) {
+		p11_message (_("no type specified"));
+		return false;
+	}
+
+	switch (type) {
+#ifdef P11_KIT_TESTABLE
+	case CKM_MOCK_GENERATE:
+		break;
+#endif
+	case CKM_RSA_PKCS_KEY_PAIR_GEN:
+		if (bits == 0) {
+			p11_message (_("no bits specified"));
+			return false;
+		}
+		break;
+	case CKM_ECDSA_KEY_PAIR_GEN:
+	case CKM_EC_EDWARDS_KEY_PAIR_GEN:
+		if (ec_params->data == NULL) {
+			p11_message (_("no curve specified"));
+			return false;
+		}
+		break;
+	default:
+		p11_message (_("unkwnown mechanism type in %s"), __func__);
+		return false;
+	}
+
+	if (bits != 0 && ec_params->data != NULL) {
+		p11_message (_("both %s and %s cannot be specified at once"), "--bits", "--curve");
+		return false;
+	}
+
+	return true;
+}
+
+static bool
+get_templates (const char *label,
+	       CK_MECHANISM_TYPE type,
+	       const CK_ULONG *bits,
+	       const p11_buffer *ec_params,
+	       CK_ATTRIBUTE **pubkey,
+	       CK_ATTRIBUTE **privkey)
+{
+	CK_BBOOL tval = CK_TRUE, fval = CK_FALSE;
+	CK_ATTRIBUTE *pub = NULL, *priv = NULL, *tmp;
+	CK_ATTRIBUTE attr_token = { CKA_TOKEN, &tval, sizeof (tval) };
+	CK_ATTRIBUTE attr_private_true = { CKA_PRIVATE, &tval, sizeof (tval) };
+	CK_ATTRIBUTE attr_private_false = { CKA_PRIVATE, &fval, sizeof (fval) };
+	CK_ATTRIBUTE attr_sign = { CKA_SIGN, &tval, sizeof (tval) };
+	CK_ATTRIBUTE attr_verify = { CKA_VERIFY, &tval, sizeof (tval) };
+	CK_ATTRIBUTE attr_decrypt = { CKA_DECRYPT, &tval, sizeof (tval) };
+	CK_ATTRIBUTE attr_encrypt = { CKA_ENCRYPT, &tval, sizeof (tval) };
+	CK_ATTRIBUTE attr_modulus = { CKA_MODULUS_BITS, (void *)bits, sizeof (*bits) };
+	CK_ATTRIBUTE attr_ec_params = { CKA_EC_PARAMS, (void *)bits, sizeof (*bits) };
+
+	tmp = p11_attrs_build (pub, &attr_token, &attr_private_false, &attr_sign, &attr_verify, NULL);
+	if (tmp == NULL) {
```

Comment:
This is the place you could use assertion macros like `return_val_if_fail`, without doing any cleanup, as allocation failure is fatal and the program cannot proceed.

Reply:
I think I prefer doing the cleanup anyway.